packages/core-app-api/src/app/AppManager.tsx

```diff
@@ -181,9 +178,15 @@ export class AppManager implements BackstageApp {
   private readonly apiFactoryRegistry: ApiFactoryRegistry;
 
   constructor(options: AppOptions) {
+    this.pluginMetadataExtender = new PluginMetadataExtender(options);
     this.apis = options.apis ?? [];
     this.icons = options.icons;
-    this.plugins = new Set((options.plugins as CompatiblePlugin[]) ?? []);
+    this.plugins = new Set(
+      ((options.plugins as CompatiblePlugin[]) ?? []).map(plugin => {
+        plugin.setMetadataExtender?.(this.pluginMetadataExtender.extend);
+        return plugin;
+      }),
```

Comment:
It's complicated to get away from. As said, we need to allow `setMetadataExtender` to not exist in older version of plugins, and plugins need to access metadata when doing `getInfo()`, but the metadata is injected into the _app_.

So, metadata needs to be provided from the app to the plugins, but the metadata retrieval functions are async and best invoked only when needed (on `getInfo()` calls) to prevent unnecessary dyn-loading of metadata chunks.

Hence this optional _"call this function if you need to get metadata"_ setter...

Reply:
Tbh I was kinda just thinking that we should split the iteration out into a separate `.forEach`, with the plugins assignment being separate 😅